File: src/game_fetcher.erl
-module(game_fetcher).
-export([start/0, test_api/0]).

-define(API_KEY, "...").
-define(BASE_URL, "https://na1.api.riotgames.com").
-define(MAX_PLAYERS_TO_CHECK, 200).
-define(DELAY_BETWEEN_CHECKS, 500).

start() ->
    case application:ensure_all_started([jsx, inets, ssl]) of
        {ok, _} ->
            test_api();
        {error, Reason} ->
            io:format("Failed to start required applications: ~p~n", [Reason])
    end.

test_api() ->
    Url = ?BASE_URL ++ "/lol/league/v4/entries/RANKED_SOLO_5x5/IRON/IV",
    Headers = [{"X-Riot-Token", ?API_KEY}],
    
    case httpc:request(get, {Url, Headers}, [], []) of
        {ok, {{_, 200, _}, _, Body}} ->
            Decoded = jsx:decode(list_to_binary(Body)),
            case Decoded of
                [] -> io:format("No Iron players found~n");
                Players -> 
                    Shuffled = shuffle_list(Players),
                    check_players(Shuffled, 0)
            end;
        {ok, {{_, Code, _}, _, _}} ->
            io:format("API error getting Iron players. Status code: ~p~n", [Code]);
        Error ->
            io:format("Network error getting Iron players: ~p~n", [Error])
    end.

shuffle_list(List) ->
    Numbered = lists:zip(lists:seq(1, length(List)), List),
    Shuffled = lists:sort([{rand:uniform(), N, E} || {N, E} <- Numbered]),
    [E || {_, _, E} <- Shuffled].

check_players([], _Count) ->
    io:format("No active games found~n");
check_players(_Players, Count) when Count >= ?MAX_PLAYERS_TO_CHECK ->
    io:format("Checked ~p players, no active games found~n", [Count]);
check_players([Player|Rest], Count) ->
    io:format("\rChecking player ~p/~p...", [Count + 1, ?MAX_PLAYERS_TO_CHECK]),
    
    case check_player_game(maps:get(<<"summonerId">>, Player)) of
        {found, GameInfo} ->
            io:format("~nFound active game! Attempting to connect...~n"),
            connect_to_game(GameInfo);
        not_found ->
            timer:sleep(?DELAY_BETWEEN_CHECKS),
            check_players(Rest, Count + 1);
        {error, Reason} ->
            io:format("~nError checking player: ~p~n", [Reason]),
            check_players(Rest, Count + 1)
    end.

check_player_game(SummonerId) ->
    case get_summoner_info(SummonerId) of
        {ok, PUUID} ->
            check_active_game(PUUID);
        {error, Reason} ->
            {error, {summoner_info_failed, Reason}}
    end.

get_summoner_info(SummonerId) ->
    Url = ?BASE_URL ++ "/lol/summoner/v4/summoners/" ++ uri_string:quote(binary_to_list(SummonerId)),
    Headers = [{"X-Riot-Token", ?API_KEY}],
    
    case httpc:request(get, {Url, Headers}, [], []) of
        {ok, {{_, 200, _}, _, Body}} ->
            SummonerInfo = jsx:decode(list_to_binary(Body)),
            {ok, maps:get(<<"puuid">>, SummonerInfo)};
        {ok, {{_, Code, _}, _, _}} ->
            {error, {api_error, Code}};
        Error ->
            {error, {network_error, Error}}
    end.

check_active_game(PUUID) ->
    Url = ?BASE_URL ++ "/lol/spectator/v5/active-games/by-summoner/" ++ uri_string:quote(binary_to_list(PUUID)),
    Headers = [{"X-Riot-Token", ?API_KEY}],
    
    case httpc:request(get, {Url, Headers}, [], []) of
        {ok, {{_, 200, _}, _, Body}} ->
            try
                GameData = jsx:decode(list_to_binary(Body)),
                GameId = maps:get(<<"gameId">>, GameData),
                EncKey = maps:get(<<"encryptionKey">>, maps:get(<<"observers">>, GameData)),
                {found, {GameId, EncKey}}
            catch
                error:{badkey, Key} ->
                    {error, {missing_key, Key}};
                Error:Reason ->
                    {error, {json_parse_error, {Error, Reason}}}
            end;
        {ok, {{_, 404, _}, _, _}} ->
            not_found;
        {ok, {{_, Code, _}, _, _}} ->
            {error, {api_error, Code}};
        Error ->
            {error, {network_error, Error}}
    end.

connect_to_game({GameId, EncryptionKey}) ->
    % Here we'll need to:
    % 1. Connect to spectator service
    % 2. Use GameId and EncryptionKey to authenticate
    % 3. Start receiving game data
    io:format("Game ID: ~p~nEncryption Key: ~p~n", [GameId, EncryptionKey]),
    % TODO: Implement actual spectator connection
    {ok, {GameId, EncryptionKey}}.

File: WoodTierTavern/src/gameFetcher.js
// gameFetcher.js
const { LeagueClient } = require('./leagueClient');
const { getApiKey, delay, readPlayerFile, shuffleList, updateProgress, makeRiotRequest } = require('./utils');

class GameFetcher {
    constructor() {
        this.API_KEY = null;
        this.FILE_PATH = "../files/";
        this.MIN_GAME_DURATION = 900;
        this.DELAY_BETWEEN_CHECKS = 1350;
        this.BASE_URL = 'https://na1.api.riotgames.com';
        this.AMERICAS_URL = 'https://americas.api.riotgames.com';
        this.IRON_FILES = [
            `${this.FILE_PATH}iron_i_players.json`,
            `${this.FILE_PATH}iron_ii_players.json`,
            `${this.FILE_PATH}iron_iii_players.json`,
            `${this.FILE_PATH}iron_iv_players.json`
        ];
    }

    async checkActiveGame(player) {
        const url = `${this.BASE_URL}/lol/spectator/v5/active-games/by-summoner/${encodeURIComponent(player.puuid)}`;
        await delay(this.DELAY_BETWEEN_CHECKS);
    
        try {
            const response = await makeRiotRequest(this.API_KEY, url);
            const rankedQueueIds = [420];
    
            if (rankedQueueIds.includes(response.gameQueueConfigId)) {
                const participant = response.participants.find(p => p.puuid === player.puuid);
                if (!participant) {
                    return { found: false };
                }
    
                if (response.gameLength >= this.MIN_GAME_DURATION) {
                    return {
                        found: true,
                        gameId: response.gameId,
                        encryptionKey: response.observers.encryptionKey,
                        gameQueueConfigId: response.gameQueueConfigId,
                        summonerId: player.summonerId,
                        puuid: player.puuid,
                        gameName: player.gameName,
                        tagLine: player.tagLine,
                        gameLength: response.gameLength
                    };
                } else {
                    console.log(`\nGame found but has been running for ${Math.floor(response.gameLength / 60)}:${(response.gameLength % 60).toString().padStart(2, '0')} minutes`);
                    return { found: false };
                }
            }
            return { found: false };
        } catch (error) {
            if (error.response?.status === 404) {
                return { found: false };
            }
            throw error;
        }
    }

    async checkPlayers(players, count) {
        if (players.length === 0) {
            console.log('\nFinished checking all players. No active games found');
            return;
        }

        const player = players[0];
        updateProgress(count + 1, players.length, 'Checking player');

        try {
            const gameInfo = await this.checkActiveGame(player);

            if (gameInfo.found) {
                const minutes = Math.floor(gameInfo.gameLength / 60);
                const seconds = gameInfo.gameLength % 60;
                
                console.log(`\nFound active game! Game running for ${minutes}:${seconds.toString().padStart(2, '0')}. Attempting to spectate...`);
                
                const leagueClient = new LeagueClient();
                try {
                    await leagueClient.startSpectate(gameInfo);
                    return true;
                } catch (error) {
                    console.error('Failed to spectate game:', error.message);
                    return false;
                }
            } else {
                return await this.checkPlayers(players.slice(1), count + 1);
            }
        } catch (error) {
            console.error('\nError checking player:', error);
            return await this.checkPlayers(players.slice(1), count + 1);
        }
    }

    async start() {
        this.API_KEY = await getApiKey();
    
        const allPlayers = [];
        for (const file of this.IRON_FILES) {
            const players = await readPlayerFile(file);
            allPlayers.push(...players);
        }
    
        if (allPlayers.length === 0) {
            console.log('No players found in files');
            return;
        }
    
        const shuffledPlayers = shuffleList(allPlayers);
        const result = await this.checkPlayers(shuffledPlayers, 0);
        return result;
    }
}

module.exports = { GameFetcher };